zds/tutorialv2/views/published.py

```diff
@@ -87,31 +87,40 @@ def get_context_data(self, **kwargs):
             .filter(related_content__pk=self.object.pk) \
             .order_by('pubdate')
 
-        # pagination of articles
-        context['paginate_articles'] = False
-
-        if self.object.type == 'ARTICLE':
-            # fetch all articles in order to find the previous and the next one
-            all_articles = \
-                [a for a in PublishedContent.objects
-                    .filter(content_type='ARTICLE', must_redirect=False)
-                    .order_by('publication_date')
-                    .all()]
-            articles_count = len(all_articles)
+        # pagination of articles and opinions
+        context['previous_content'] = None
+        context['next_content'] = None
+
+        if self.current_content_type in ('ARTICLE', 'OPINION'):
+            # fetch all articles or opinions in order to find the previous and the next one
+            queryset_pagination = PublishedContent.objects
+
+            if self.current_content_type == 'OPINION':
+                # filter opinions only from the same author
+                queryset_pagination = queryset_pagination.filter(content_type=self.current_content_type,
+                                                                 must_redirect=False,
+                                                                 authors__in=self.object.authors.all())
+            else:
+                queryset_pagination = queryset_pagination.filter(content_type=self.current_content_type,
+                                                                 must_redirect=False)
+
+            all_content = []
+            for c in queryset_pagination.order_by('publication_date').all():
+                if c not in all_content:
+                    # we don't like duplicated entries
+                    all_content.append(c)
```

Comment:
Je ne comprend pas trop le pourquoi de ces quelques lignes. Si le but est juste d'obtenir des objets uniques, il est possible d'utiliser [distinct](https://docs.djangoproject.com/fr/1.11/ref/models/querysets/#django.db.models.query.QuerySet.distinct).

Sinon de façon générale : 
- En python le set sont pratiques pour l'unicité des éléments
- On peut réaliser ce genre de code de façon plus immédiate en utilisant une list compréhension

Reply:
distinct ne fonctionne pas avec SQLite donc je ne l'ai pas utilisé (impossibilité de tester en local c'est très bof) mais il aurait été très utile ici !

J'ai préféré écrire sur quelques lignes pour que ce soit plus lisible qu'une liste de compréhension, mais je ne connaissais pas le set donc je vais arranger ça :)

Reply:
> distinct ne fonctionne pas avec SQLite

Alors la je suis très étonné, j'ai fait quelques recherche et je n'ai rien trouvé qui affirme que ça ne fonctionne pas, par exemple:
http://www.sqlitetutorial.net/sqlite-select-distinct

Est-ce que tu as essayé de fournir un ou plusieurs fields à distinct (par exemple `.distinct(id)` ?), car il semblerait que soit utilisé comme ça en SQLite plutôt que sans expression.

Etant cette fois ci sur un PC, je développe un peu :

Si vraiment ça ne fonctionne pas tu as pour moi deux façon de faire:

List comprehension
```py
qs = queryset_pagination.order_by('publication_date').all()
all_content = [c for c in qs if c not in all_content]
```

Mais comme dit précédemment ce n'est pas vraiment un bonne usage d'une list compréhension  car on cherche juste à enlever les doublons. Une autre solution en utilisant les sets:

```py
qs = queryset_pagination.order_by('publication_date').all()
all_content = list(set(qs))
```

Ici je convertit à nouveau en liste pour obtenir un gain de performance si on souhaite itérer dessus par la suite. Si on souhait juste vérifier de l'appartenance, un Set est plus rapide. Je pense que dans ce cas tu sera obligé de convertir  nouveau en liste.

Reply:
Par contre euh… Je connais pas très bien le backend mais actuellement, le `for`, dans le cas d’un article, il itère bien sur *tous* les articles de la BDD, non ? Ça risque pas d’être un peu lent ça ?

’fin pour moi, la bonne solution est ni d’utiliser un `set`, ni une compréhension ni rien, mais de faire une query pour obtenir uniquement la publication prédence et une autre pour obtenir uniquement la publication suivante, histoire que la lenteur de ce truc ne soit pas proportionelle au nombre de publications. Ou alors d’optimiser ça en une seule query si possible, mais sans récupérer *toutes* les publications si vous voyez ce que je veut dire.

Reply:
@motet-a : C'est vrai. En fait j'était passé dessus uniquement d'un point de vu stylistique, mais en effet.

D'après ce que j'ai compris cette requête sert à 3 choses:
- Compter le nombre de contenus (avec len() )
- Avoir le contenu précédent
- Avoir le contenu suivant

Il est vrai que choper toutes les publications pour ça parait un peu abusé. Je propose en code corrigé la version suivante (non testé):


```py
if self.current_content_type in ('ARTICLE', 'OPINION'):
    queryset_pagination = PublishedContent.objects.filter(content_type=self.current_content_type, must_redirect=False)

    if self.current_content_type == 'OPINION':
        # filter opinions only from the same author
        # Note: j'ai ajouté le distinct uniquement ici car il me semble que la duplication ne peut que venir du authors__in
        queryset_pagination = queryset_pagination.filter(authors__in=self.object.authors.all()).dictinct('id')

    contents_count = queryset_pagination.count()
    previous_content = queryset_pagination.filter(id__lt=self.public_content_object.id).order_by('id').first()
    next_content = queryset_pagination.filter(id__gt=self.public_content_object.id).order_by('id').first()
```

Je pense que motet-a tu avait quelque chose comme ça en tête. Ça devrait être plus rapide et scaler plus rapidement, même si on fait plus de requêtes elles sont bien plus simples. Ca serait cool de faire un petit benchmark pour comparer.

Reply:
Ouais sauf que je pense qu’on peut enlever le `contents_count`.

Reply:
Concernant les benchmarks, je ne pense pas que ce soit la peine. Un rapide coup d’œil sur le SQL généré par ce bazard devrait être suffisant.

Reply:
Je retesterai avec le mot clé DISTINCT :)

Effectivement itéré sur tous les articles ça fait une requête très lourde et il y a probablement mieux. Par contre je ne sais pas si la solution de @Anto59290 fonctionnera telle quelle car on ne veut pas trier en fonction de `id` mais du champ `publication_date` qui est une date et pas un entier, mais je vais faire des recherches pour essayer de trouver une solution. :)

Reply:
En fait c’est quoi le problème avec le tri sur la date ? Toutes les grosses bases de données gèrent ça très bien.

Reply:
Je dis juste que je ne sais pas si ça fonctionne ou pas, mais si ça fonctionne alors ok :)

Reply:
Ça fonctionne forcément :) Enfin, faudra peut-être rajouter un index, je sais pas, j’y connais pas grand-chose. Mais on peut forcément dire à MySQL de trier par date, c’est un classique.

Reply:
> Ouais sauf que je pense qu’on peut enlever le contents_count.

En effet, encore mieux du coup

> Concernant les benchmarks, je ne pense pas que ce soit la peine. 

C'était surtout a but pédagogique. Si je trouve un peu de temps cette semaine j'en ferais avec plusieurs contenus

Normalement ça fonctionnera en effet quelque soit le type du champ. Pour l'index faudrait voir, mais en effet mettre un index sur la date de publication serait pas forcément con. Ce n'est pas vraiment en rapport avec cette PR cela dit.